### DISCUSSION TITLE
Decouple changelog retrieval from datasources

### DISCUSSION THREAD

**Discussion by secustor** [renovatebot/renovate#10992] [Suggest an Idea]
As there are more changelog specific issues latley, at least it seems to me.  I habe been thinking about the changelog retrieval.  
One of the scenarios which I have seen is to use dockerregistry images which have been replicated into a local registry as datasource and Gitlab.com as source of the changelog.  As I'm aware this is not possible to handle right now.  
 
My proposal is to move the changelog logic into classes and handle them like versioning is handled for datasources right now. That would allow users to overwrite the sources based on type and later template them maybe with a generic approach. 

WDYT?

**Comment by rarkins** [comment]
I support a rewrite of changelog handling and converting them to more like "modules" like we do with others, but I don't think changelogs are really coupled with datasources today - they are semi-coupled with platforms though (github and gitlab). Datasources can return `sourceUrl` and `changelogUrl` values but after that all fetching is done fairly independently.

Some coupling with datasource occurs when we try to determine "which releases occurred between X and Y?". Perhaps that's something the datasource could calculate and include in the lookup result, but it doesn't make a huge difference.

Something we don't support today but which would be nice is to support a "generic" changelog URL parsing, so that it doesn't have to be hosted on a code platform.

We also struggle when there is both "GitHub Releases" as well as a CHANGELOG.md file in the repo. Sometimes they duplicate the data, sometimes they have empty releases and the real info is in the changelog file.
